Portable SIMD subtree update

dpaoliello opened the pull request:
Syncs nightly to the latest changes from rust-lang/portable-simd

r? @calebzulawski

taiki-e committed:
Fix build error on big endian aarch64

taiki-e committed:
Remove redundant imports

calebzulawski committed:
Merge pull request #397 from taiki-e/imports

Remove redundant imports

calebzulawski committed:
Merge pull request #396 from taiki-e/aarch64-big

Fix build error on big endian aarch64

dpaoliello committed:
Add arm64ec support

AquaEBM committed:
add stdarch_x86_avx512 feature flag for AVX-512-supporting architectures

calebzulawski committed:
Merge pull request #398 from AquaEBM/avx512-feature-flag

fix build error on AVX-512

farnoy committed:
Add support for masked loads & stores

Commit:
feat: add SIMD float math functions (exp, exp2, log, log2, log10, sin, cos).

calebzulawski committed:
Merge pull request #400 from avhz/master

feat: add SIMD float math functions (exp, exp2, log, log2, log10, sin…

calebzulawski committed:
Test std_float

calebzulawski committed:
Update docs

calebzulawski committed:
Fix formatting

calebzulawski committed:
Fix wasm tests

calebzulawski committed:
Attempt to avoid LLVM error

calebzulawski committed:
Merge pull request #401 from rust-lang/std_float_improvements

Test StdFloat

calebzulawski committed:
Merge pull request #404 from dpaoliello/arm64ec

Add arm64ec support

calebzulawski committed:
Merge pull request #399 from farnoy/masked-load-store-simple

Add support for masked loads & stores

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8396109680/job/22996802794) [(plain)](https://github.com/rust-lang/rust/commit/86a391c3f0376e9d36bb0cd60945ac1ef1ea6e61/checks/22996802794/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................

failures:

---- library/core/src/../../portable-simd/crates/core_simd/src/vector.rs - core_simd::vector::Simd<T,N>::load_or (line 347) stdout ----
error: unused import: `Mask`
error: doctest failed, to rerun pass `-p core --doc`
  |
  |
7 | use simd::{Simd, Mask};
  |
note: the lint level is defined here
 --> library/core/src/../../portable-simd/crates/core_simd/src/vector.rs:345:9
  |
---
---- library/core/src/../../portable-simd/crates/core_simd/src/vector.rs - core_simd::vector::Simd<T,N>::load_or_default (line 324) stdout ----
error: unused import: `Mask`
##[error] --> library/core/src/../../portable-simd/crates/core_simd/src/vector.rs:328:18
  |
7 | use simd::{Simd, Mask};
  |
note: the lint level is defined here
 --> library/core/src/../../portable-simd/crates/core_simd/src/vector.rs:322:9
  |
```

</details>

calebzulawski commented:
Can you combine your two additional commits?

dpaoliello commented:
> Can you combine your two additional commits?

Done

workingjubilee commented on library/portable-simd/crates/core_simd/src/vector.rs:
```diff
@@ -1,3 +1,4 @@
+use crate::convert::{TryFrom, TryInto};
```
This... should not be necessary?

dpaoliello replied:
I saw this when building locally, I assume it's because the prelude is not automatically imported when building the `core` crate itself?

```
   Compiling core v0.0.0 (rust\library\core)
error[E0599]: no method named `try_into` found for reference `&[T]` in the current scope
    --> library\core\src\..\..\portable-simd\crates\core_simd\src\vector.rs:1051:35
     |
1051 |         Ok(Self::from_array(slice.try_into()?))
     |                                   ^^^^^^^^
     |
     = help: items from traits can only be used if the trait is in scope
     = note: 'crate::convert::TryInto' is included in the prelude starting in Edition 2021
help: trait `TryInto` which provides `try_into` is implemented but not in scope; perhaps you want to import it
```

dpaoliello replied:
Yep, `#![no_core]` crates, like `core` itself https://github.com/rust-lang/rust/blob/b3df0d7e5ef5f7dbeeca3fb289c65680ad013f87/library/core/src/lib.rs#L95
Do not get a prelude injected https://github.com/rust-lang/rust/blob/b3df0d7e5ef5f7dbeeca3fb289c65680ad013f87/compiler/rustc_builtin_macros/src/standard_library_imports.rs#L23-L24

workingjubilee replied:
That cannot be correct, consider this code in `core` with no `use crate::option::Option::Some` in sight:

https://github.com/rust-lang/rust/blob/b3df0d7e5ef5f7dbeeca3fb289c65680ad013f87/library/core/src/slice/mod.rs#L104-L115

workingjubilee replied:
@dpaoliello Core is far more demonic than that, it is simultaneously exempted from having a prelude and **deliberately** imports one using compiler magic:

https://github.com/rust-lang/rust/blob/b3df0d7e5ef5f7dbeeca3fb289c65680ad013f87/library/core/src/lib.rs#L285-L287

workingjubilee replied:
Ah, I see my confusion now, we only import the v1 prelude.

dpaoliello replied:
Ah, that'd do it, the `Try*` traits are in the Rust 2021 prelude, not the v1 prelude.

workingjubilee replied:
I bet in the portable-simd repo that landing this import causes a lint about redundant imports tho'.

dpaoliello replied:
Yep, that's likely. I believe the correct fix would be to have `core` import the 2021 prelude, but that's a much larger change than I'm willing to do here...

dpaoliello replied:
Happy to make that change, if that's what the lib team wants, just not as a part of this PR.

dpaoliello replied:
Filed https://github.com/rust-lang/rust/issues/122912

calebzulawski replied:
We can also modify the prelude imported by the portable-simd crate to make it act like core.

dpaoliello replied:
Let's see what happens with that issue first: I'd rather not add a workaround to portable-simd if we can get core to do the "right thing" instead.

workingjubilee replied:
> Yep, that's likely. I believe the correct fix would be to have `core` import the 2021 prelude, but that's a much larger change than I'm willing to do here...

yeaaaah let's not dig into that too deeply. ^^;

calebzulawski replied:
I opened https://github.com/rust-lang/portable-simd/pull/406 to avoid issues when we sync changes back.

workingjubilee reviewed (commented):
I apologize, I believe the rust-lang/rust to rust-lang/portable-simd direction should happen first? I believe that is the only reason this is breaking tests.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8396486643/job/22997984288) [(plain)](https://github.com/rust-lang/rust/commit/08bd6e0839f3499e7277fbfc9efd649b0e772462/checks/22997984288/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling compiler_builtins v0.1.108
error: unresolved link to `core::ptr::wrapping_offset`
   --> library/core/src/../../portable-simd/crates/core_simd/src/vector.rs:698:55
    |
698 |     /// Memory addresses for element are calculated [`core::ptr::wrapping_offset`] and
    |                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^ no item named `wrapping_offset` in module `ptr`
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`

error: could not document `core`
```

</details>

workingjubilee commented:
`wrapping_offset` is an implemented fn on the primitive type `pointer`, not a function in the module `core::ptr`.

workingjubilee commented on library/portable-simd/crates/core_simd/src/vector.rs:
```diff
@@ -493,6 +632,77 @@ where
+    ///
+    /// # Safety
+    ///
+    /// Memory addresses for element are calculated [`core::ptr::wrapping_offset`] and
+    /// each enabled element must satisfy the same conditions as [`core::ptr::write`].
```
`wrapping_offset` is an implemented fn on the primitive type pointer, not a function in the module `core::ptr`.

dpaoliello commented:
> I apologize, I believe the rust-lang/rust to rust-lang/portable-simd direction should happen first? I believe that is the only reason this is breaking tests.

There were no changes in the rust-lang/rust `portable-simd`. The doc failures were added 2 weeks ago (https://github.com/rust-lang/portable-simd/commit/4f0ba1ae1963cf785370f0b951018ed214dc48bf), and the `Try*` traits were previously imported, but then removed in a cleanup last month (https://github.com/rust-lang/portable-simd/commit/bb4bba5fcf9fce9208e6c2c1de008cec78a9519c)

workingjubilee commented:
oh okay!

calebzulawski committed:
Use v1 prelude to match core (rust-lang/portable-simd#406)

dpaoliello committed:
Validate generating docs (rust-lang/portable-simd#407)

- Generate docs with warnings disabled.
- Instruct doctest to add `#![deny(warnings)]` to all tests.

dpaoliello commented:
Any further changes requested? Otherwise, can someone please put this in the queue?

workingjubilee commented:
Ah, is there any chance you could make it so this was based on portable-simd tip-of-tree instead of having 6736d7b5a3e4247326312dda106ab8d741c6cfdf be an empty commit when synced back? ^^;

workingjubilee commented:
This should go in alone when it does go.

@bors rollup=never p=1

dpaoliello committed:
Merge commit 'cff979eec1ac0473fc4960ee6cde462c6aeda824' into sync-portable-simd-2024-03-22

dpaoliello commented:
> Ah, is there any chance you could make it so this was based on portable-simd tip-of-tree instead of having [6736d7b](https://github.com/rust-lang/rust/commit/6736d7b5a3e4247326312dda106ab8d741c6cfdf) be an empty commit when synced back? ^^;

Done

workingjubilee commented:
Thanks! Sorry about that, I know this gets to be a huge pain in the ass. ^^;
In we go!

@bors r+

bors commented:
:pushpin: Commit 9e0ec251d53010ebe6f5c052abc597c0666480a5 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 9e0ec251d53010ebe6f5c052abc597c0666480a5 -->
<!-- homu: {"type":"Approved","sha":"9e0ec251d53010ebe6f5c052abc597c0666480a5","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

RalfJung commented on library/portable-simd/crates/std_float/src/lib.rs:
```diff
@@ -101,46 +143,65 @@ pub trait StdFloat: Sealed + Sized {
-        let _ = x2.abs() * x2;
-    }
+impl_float! {
+    sin: simd_fsin,
```
Looks like the existing `sqrt` could also be generated with this macro now?

calebzulawski replied:
It's not necessary, this macro is avoiding an ISel bug on aarch64 f64x1 that doesn't happen with sqrt.

RalfJung replied:
Ah, I thought it was meant to just make this a bit shorter.^^ Never mind then.

bors commented:
:hourglass: Testing commit 9e0ec251d53010ebe6f5c052abc597c0666480a5 with merge 9b8d12cf4c2311203aea83315552b15993bd4f81...
<!-- homu: {"type":"BuildStarted","head_sha":"9e0ec251d53010ebe6f5c052abc597c0666480a5","merge_sha":"9b8d12cf4c2311203aea83315552b15993bd4f81"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8405347243/job/23019647811)
Approved by: workingjubilee
Pushing 9b8d12cf4c2311203aea83315552b15993bd4f81 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"workingjubilee","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8405347243/job/23019647811"},"merge_sha":"9b8d12cf4c2311203aea83315552b15993bd4f81"} -->

rust-timer commented:
Finished benchmarking commit (9b8d12cf4c2311203aea83315552b15993bd4f81): [comparison URL](https://perf.rust-lang.org/compare.html?start=2f090c30ddd6b3bbe5c81c087579a5166e7c7278&end=9b8d12cf4c2311203aea83315552b15993bd4f81&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=2f090c30ddd6b3bbe5c81c087579a5166e7c7278&end=9b8d12cf4c2311203aea83315552b15993bd4f81&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=2f090c30ddd6b3bbe5c81c087579a5166e7c7278&end=9b8d12cf4c2311203aea83315552b15993bd4f81&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=2f090c30ddd6b3bbe5c81c087579a5166e7c7278&end=9b8d12cf4c2311203aea83315552b15993bd4f81&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=2f090c30ddd6b3bbe5c81c087579a5166e7c7278&end=9b8d12cf4c2311203aea83315552b15993bd4f81&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 669.644s -> 669.289s (-0.05%)
**Artifact size**: 314.97 MiB -> 315.03 MiB (0.02%)
